### PR TITLE
Support GH-team based policy ownership

### PR DESCRIPTION
**Pull request by samrabelachew**
Current policy check ownership logic relies on validating owner exists in a global list field defined in the startup configuration. Instead, this allows for a more granular approach to policy approvals by allowing specific GH team members to approve a policy specifically owned by them.

Changes include:
* creating an atlantis config at the policy level that defines the team name that "owns" the policy
* relying on above config + GH teams api to identify the current set of team members that owns the policy
* within the policy filter logic, we use check if a failing policy in a PR is already approved by a policy owner and filter those out when determining of the atlantis policy check run should fail
* to avoid constantly calling the GH teams API, we keep an in-memory cache of policy owners

### CODE REVIEW

File: server/core/config/raw/policies.go [hunk 1]
```diff
@@ -52,16 +52,16 @@ func (o PolicyOwners) ToValid() valid.PolicyOwners {
...
 
 type PolicySet struct {
-	Path   string       `yaml:"path" json:"path"`
-	Source string       `yaml:"source" json:"source"`
-	Name   string       `yaml:"name" json:"name"`
-	Owners PolicyOwners `yaml:"owners,omitempty" json:"owners,omitempty"`
+	Path   string `yaml:"path" json:"path"`
+	Source string `yaml:"source" json:"source"`
+	Name   string `yaml:"name" json:"name"`
+	Owner  string `yaml:"owner,omitempty" json:"owner,omitempty"`
```
Comment (asubrocky) [inline server/core/config/raw/policies.go:58]: Was this "owners" field used previously?

File: server/vcs/provider/github/team_fetcher.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+	"github.com/palantir/go-githubapp/githubapp"
+	"github.com/pkg/errors"
+)
+
+type TeamMemberFetcher struct {
+	ClientCreator githubapp.ClientCreator
+	Org           string
+}
+
+func (t *TeamMemberFetcher) ListTeamMembers(ctx context.Context, installationToken int64, teamSlug string) ([]string, error) {
```
Comment (asubrocky) [inline server/vcs/provider/github/team_fetcher.go:15]: Have some potentially naive questions here 😄 

* How are these github teams provisioned?  Is it with metaservice?  
* Have we evaluated using something like the internal acl/teams service?

**Review by asubrocky — COMMENTED** [review]
Looks good overall, just had a few questions to help me gain understanding

File: server/core/config/raw/policies.go (hunk 1 above)
Comment (samrabelachew) [reply server/core/config/raw/policies.go:58]: No, it wasn't used yet.

File: server/vcs/provider/github/team_fetcher.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+	if err != nil {
+		return nil, errors.Wrap(err, "creating installation client")
+	}
+	var usernames []string
+	run := func(ctx context.Context, nextPage int) ([]*gh.User, *gh.Response, error) {
+		listOptions := gh.ListOptions{
+			PerPage: 100,
+		}
+		listOptions.Page = nextPage
+		return client.Teams.ListTeamMembersBySlug(ctx, t.Org, teamSlug, &gh.TeamListTeamMembersOptions{})
```
Comment (Aayyush) [inline server/vcs/provider/github/team_fetcher.go:26]: nit: pass the listOptions to the method call

**Review by Aayyush — APPROVED** [review]
left a minor nit! Do we plan to filter out a policy set if an owner is the PR author?

File: server/vcs/provider/github/team_fetcher.go (hunk 3 above)
Comment (samrabelachew) [reply server/vcs/provider/github/team_fetcher.go]: fixed!

**Comment by samrabelachew** [comment]
> left a minor nit! Do we plan to filter out a policy set if an owner is the PR author?

I believe it shouldn't matter if we do since we only look at who approved a policy in the policy filter logic and PR authors can't be approvers.

File: server/events/policy_filter.go [hunk 4]
```diff
@@ -11,31 +11,83 @@ type prReviewsFetcher interface {
...
 
+type teamMemberFetcher interface {
+	ListTeamMembers(ctx context.Context, installationToken int64, teamSlug string) ([]string, error)
+}
+
 type ApprovedPolicyFilter struct {
-	//TODO: remove in favor of an in-memory cache of owners
-	GlobalCfg        valid.GlobalCfg
-	PRReviewsFetcher prReviewsFetcher
+	owners map[string][]string //cache
```
Comment (nishkrishnan) [inline server/events/policy_filter.go:19]: Do you want to use sync.Map for this?  Currently this would be problematic when used across multiple go routines.

File: server/core/config/raw/policies_test.go [hunk 5]
```diff
@@ -75,17 +75,13 @@ func TestPolicySets_Validate(t *testing.T) {
...
 					{
-						Name: "policy-name-2",
-						Owners: raw.PolicyOwners{
-							Users: []string{
-								"john-doe",
-								"jane-doe",
-							},
-						},
+						Name:   "policy-name-2",
+						Owner:  "owner2",
```
Comment (nishkrishnan) [inline server/core/config/raw/policies_test.go:84]: Why was this changed to a string? The original intention of using a struct was to allow us to add teams as a field so that we didn't have to make a bunch of these code changes.
